### PR TITLE
Filter lock

### PR DESCRIPTION
**Pull request by dnil**
This PR adds a functionality: persistent filters can be locked by a user, that then becomes its owner. While the lock is active, other users cannot remove the filter. Tentatively introduce optional audit trail for filters.

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by hassanfa** [comment]
This is amazing! We were just talking about something similar to this. Looking forward to use this. 💯

**Comment by northwestwitch** [comment]
Sorry I see the grey button and my intention is to update with latest master, but it's the button to mark it ready for review 😄

**Comment by northwestwitch** [comment]
OMG the conflicts. We should focus on merging this one soon!

**Comment by dnil** [comment]
Well, it's ready for review... 😉

**Comment by northwestwitch** [comment]
> Well, it's ready for review... 😉

Ah I didn't notice, I can review it tomorrow!

**Comment by northwestwitch** [comment]
Question: what does the audit filter do exactly?

**Comment by dnil** [comment]
It leaves a voluntary log trail to answer questions like "did I really check for recessive variants" or "did we do bothe the stringent and relaxed filter on this tumor"?

The idea is that the operator loads a set filter, checks through the variants and then ticks off an audit. I considered making it like a modal with the option to comment, but most can be expressed in filter terms and name, and these type of case comments are common anyway (like "no plausible SVs found with the NMD panel").

**Review by northwestwitch — APPROVED** [review]
Really nice functionalities! Nice the buttons with the javascript controls in the variants forms! I've kind of understood the audit thing and I can really see its advantages. 

The idea is to lock a filter so it can be used in the audits. Other users won't be able to modify or delete it, but the filter might still change if the owner modifies it in a second moment. So I think it in the case report it should be written somewhere what the settings os the filter are. Also to give the user that prints the report for the referral to double check that those were actually the settings used when they originally checked the data.

Other than that I have some comments that are easy to address. Approving since it is working nicely!

File: scout/adapter/mongo/event.py [hunk 1]
```diff
@@ -153,6 +155,9 @@ def events(
                 if comments:
                     query["verb"] = "comment"
 
+                if audits:
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/adapter/mongo/event.py:158]: Indeed I could! I also now know exactly what kind of anonymous reviewer would be: the "authors could also cite Recent work by Yours truly" kind! 😉

[the remark this answers is not in the document]
Comment (dnil) [reply scout/adapter/mongo/event.py:158]: The category arument for that one is a little funny, given the name of the function - but now I will shut up and fix this instead.. 😁

File: scout/adapter/mongo/filter.py [hunk 2]
```diff
@@ -106,6 +106,99 @@ def stash_filter(
 
         return filter_id
 
+    def audit_filter(self, filter_id, institute_obj, case_obj, user_obj, category="snv", link=None):
+        """Mark audit of filter for case in events.
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/adapter/mongo/filter.py:110]: I'll write a few lines in docs as well!

File: scout/adapter/mongo/event.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/adapter/mongo/event.py:158]: I was being modest 😄

File: scout/adapter/mongo/filter.py [hunk 3]
```diff
@@ -106,6 +106,99 @@ def stash_filter(
...
+            filter_obj(ReturnDocument)
+        """
+        filter_obj = None
+        LOG.debug("Retrieve filter {}".format(filter_id))
+        filter_obj = self.filter_collection.find_one({"_id": ObjectId(filter_id)})
+        if filter_obj is None:
+            return
+
+        # link e.g. to the variants view where filter was created
+        if link is None:
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/adapter/mongo/filter.py:129]: I think you got an idea now, right? I know I missed this function sometimes when analysing cases; especially to remember - and say to the doctor on the other end - that, yeah, I have checked for De novo NMD, Rare recessive NMD panel and HPO AD variants without having to write that on the case page in as many words. It also fits very well with the current cancer use case.

File: scout/server/blueprints/cases/templates/cases/case_report.html [hunk 4]
```diff
@@ -212,15 +212,36 @@ <h4>Scout case analysis report</h4> - created on:&nbsp;<strong>{{report_created_
...
+          <td colspan=2>
+            <table id="audit-table" class="table table-sm" style="background-color: transparent">
+              <thead>
+                <tr>
+                  <th>Filters marked audited for case</th>
+                </tr>
+              </thead>
+              <tbody>
+                <tr>
+                 {% for audit in audits %}
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/server/blueprints/cases/templates/cases/case_report.html:235]: Right, nice catch - never happened during my tests for some reason.. 😆

[the remark this answers is not in the document]
Comment (dnil) [reply scout/server/blueprints/cases/templates/cases/case_report.html:235]: As for changes to the filter, I tentatively feel that it is good that the meaning of a label can change over time. If we had infinite space I would agree to storing and printing filters. In the end perhaps we will help the users with at least a suggested incremental version option, like for gene panels, but maybe let's see that it finds use first?

File: scout/server/blueprints/variants/templates/variants/utils.html [hunk 5]
```diff
@@ -676,21 +676,46 @@ <h4 class="modal-title" id="filterNameModalLabel">Filter display name</h4>
...
     {{ modal_prompt_filter_name(form) }}
-    <div class="col-4">
+    <div class="col-3">
       {{ form.filters(class="form-control") }}
     </div>
-    <div class="btn-group col-4" role="group">
+    <div class="btn-group col-6" role="group">
       {{ form.load_filter(class="btn btn-secondary", disabled=form.filters.choices|length==0) }}
+      {{ form.audit_filter(class="btn btn-secondary bg-info", title="Note selected filter audited for this case.", disabled=form.filters.choices|length==0) }}
+      <button name="lock_filter" type="submit" value="Lock filter" class="btn btn-primary btn-sm" aria-hidden="true" title="Toggle filter lock" id="lock_filter" disabled="{{form.filters.choices|length==0}}">
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/server/blueprints/variants/templates/variants/utils.html:689]: Sure! Toggle is more to the point though.

File: scout/server/blueprints/variants/templates/variants/utils.html [hunk 6]
```diff
@@ -676,21 +676,46 @@ <h4 class="modal-title" id="filterNameModalLabel">Filter display name</h4>
...
+              <span id="stash_filter_padlock" class="fa fa-unlock"></span></button>
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/server/blueprints/variants/templates/variants/utils.html:690]: The only downside to that is that fa-lock-open is slightly wider than fa-lock, so the button has different size. But whatever, I agree it is more easily recognisable.

File: scout/server/blueprints/variants/templates/variants/utils.html [hunk 7]
```diff
@@ -676,21 +676,46 @@ <h4 class="modal-title" id="filterNameModalLabel">Filter display name</h4>
...
     </div>
     {{ modal_prompt_filter_name(form) }}
-    <div class="col-4">
+    <div class="col-3">
       {{ form.filters(class="form-control") }}
     </div>
-    <div class="btn-group col-4" role="group">
+    <div class="btn-group col-6" role="group">
       {{ form.load_filter(class="btn btn-secondary", disabled=form.filters.choices|length==0) }}
+      {{ form.audit_filter(class="btn btn-secondary bg-info", title="Note selected filter audited for this case.", disabled=form.filters.choices|length==0) }}
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/server/blueprints/variants/templates/variants/utils.html]: I'll make it so - they are more immediate, hence for delete, but having them the same makes sense!

**Comment by dnil** [comment]
Thank you for the review! 🙇 Lots of good insights, but especially butter finger/mass use proofing the report was very good. Also the filters page looks much nicer with headers.

Screenshot 2021-05-04 at 17 02 50
Screenshot 2021-05-04 at 17 03 50

File: scout/server/blueprints/variants/templates/variants/utils.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/variants/templates/variants/utils.html]: If you use the locked version of version 5 they should be the same

[the remark this answers is not in the document]
Comment (dnil) [reply scout/server/blueprints/variants/templates/variants/utils.html]: Actually the fa4 unlock looks very much like the fa5 lock-open... Hmm...

[the remark this answers is not in the document]
Comment (dnil) [reply scout/server/blueprints/variants/templates/variants/utils.html]: No, they do appear to be different sizes, even if I disable the local css and use only the web js fa6? Nvm, it's very minor.

**Comment by dnil** [comment]
I just tested a bit more on stage; seems I missed a few things for the cancer view (changes, possibly, to be kind to myself). 😊  
Will fix first, so no merging yet!

**Pull request merged**